File: src/durable_threads/risk.py
"""Deterministic task-risk classification for economic model routing."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable

RISK_LEVELS = ("R0", "R1", "R2", "R3", "R4")
_RISK_RANK = {risk: index for index, risk in enumerate(RISK_LEVELS)}

_R4 = re.compile(
    r"\b(architecture|architectural|distributed|multi[- ]region|consensus|disaster recovery|"
    r"cross[- ]service|systemic|platform migration|control plane|data plane)\b",
    re.IGNORECASE,
)
_R3 = re.compile(
    r"\b(auth|authentication|authorization|permission|payment|billing|crypto|cryptograph|"
    r"secret|credential|privacy|pii|production|database migration|schema migration|delete data|"
    r"destructive|concurren|race condition|locking|idempotency|security|vulnerab)\w*\b",
    re.IGNORECASE,
)
_R2 = re.compile(
    r"\b(api|integration|webhook|queue|cache|database|schema|migration|cross[- ]module|"
    r"multiple modules|public contract|backward compat|dependency upgrade|release)\w*\b",
    re.IGNORECASE,
)
_R0 = re.compile(
    r"\b(typo|spelling|formatting|comment|comments|readme|documentation|docs only|rename only)\b",
    re.IGNORECASE,
)
_CHANGE = re.compile(
    r"\b(add|build|change|create|debug|fix|implement|migrate|modify|refactor|remove|repair|"
    r"update)\b",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class RiskAssessment:
    risk_class: str
    reasons: tuple[str, ...]
    explicit: bool = False


def validate_risk(value: str) -> str:
    risk = value.upper().strip()
    if risk not in _RISK_RANK:
        raise ValueError(f"risk must be one of: {', '.join(RISK_LEVELS)}")
    return risk


def meets_threshold(risk: str, threshold: str) -> bool:
    return _RISK_RANK[validate_risk(risk)] >= _RISK_RANK[validate_risk(threshold)]


def classify_risk(
    *,
    objective: str,
    allowed_paths: Iterable[str] = (),
    acceptance: Iterable[str] = (),
    explicit: str | None = None,
    default: str = "R1",
) -> RiskAssessment:
    """Classify risk conservatively from bounded task text.

    The classifier is intentionally deterministic and inspectable. It is a routing
    hint, not a security boundary. Callers can always provide an explicit risk.
    """

    if explicit:
        risk = validate_risk(explicit)
        return RiskAssessment(risk, ("risk explicitly supplied by the planner",), True)

    default = validate_risk(default)
    items = [objective, *allowed_paths, *acceptance]
    context = " ".join(item.strip() for item in items if isinstance(item, str) and item.strip())

    if _R4.search(context):
        return RiskAssessment("R4", ("systemic or architecture-level signal detected",))
    if _R3.search(context):
        return RiskAssessment(
            "R3", ("critical security, data, production, or concurrency signal detected",)
        )
    if _R2.search(context):
        return RiskAssessment("R2", ("integration or public-contract signal detected",))
    if _R0.search(context) and not _CHANGE.search(context):
        return RiskAssessment("R0", ("mechanical or documentation-only signal detected",))
    code_paths = any(
        path.startswith(("src/", "app/", "packages/")) for path in allowed_paths
    )
    if _R0.search(context) and not code_paths:
        return RiskAssessment("R0", ("mechanical or documentation-only signal detected",))
    return RiskAssessment(default, (f"no higher-risk signal detected; using default {default}",))

File: packages/policy/src/risk.ts
import { RISK_LEVELS, type RiskClass } from "@durable-threads/contracts";

const RISK_RANK: Record<RiskClass, number> = {
  R0: 0,
  R1: 1,
  R2: 2,
  R3: 3,
  R4: 4,
};

const R4 = /\b(architecture|architectural|distributed|multi[- ]region|consensus|disaster recovery|cross[- ]service|systemic|platform migration|control plane|data plane)\b/i;
const R3 =
  /\b(auth|authentication|authorization|permission|payment|billing|crypto|cryptograph|secret|credential|privacy|pii|production|database migration|schema migration|delete data|destructive|concurren|race condition|locking|idempotency|security|vulnerab)\w*\b/i;
const R2 =
  /\b(api|integration|webhook|queue|cache|database|schema|migration|cross[- ]module|multiple modules|public contract|backward compat|dependency upgrade|release)\w*\b/i;
const R0 = /\b(typo|spelling|formatting|comment|comments|readme|documentation|docs only|rename only)\b/i;
const CHANGE =
  /\b(add|build|change|create|debug|fix|implement|migrate|modify|refactor|remove|repair|update)\b/i;

export interface RiskAssessment {
  riskClass: RiskClass;
  reasons: string[];
  explicit: boolean;
}

export function validateRisk(value: string): RiskClass {
  const risk = value.toUpperCase().trim();
  if (!(risk in RISK_RANK)) {
    throw new Error(`risk must be one of: ${RISK_LEVELS.join(", ")}`);
  }
  return risk as RiskClass;
}

export function meetsThreshold(risk: string, threshold: string): boolean {
  return RISK_RANK[validateRisk(risk)] >= RISK_RANK[validateRisk(threshold)];
}

export function classifyRisk(input: {
  objective: string;
  allowedPaths?: Iterable<string>;
  acceptance?: Iterable<string>;
  explicit?: string | null;
  default?: string;
}): RiskAssessment {
  if (input.explicit) {
    return {
      riskClass: validateRisk(input.explicit),
      reasons: ["risk explicitly supplied by the planner"],
      explicit: true,
    };
  }

  const defaultRisk = validateRisk(input.default ?? "R1");
  const allowedPaths = [...(input.allowedPaths ?? [])];
  const acceptance = [...(input.acceptance ?? [])];
  const items = [input.objective, ...allowedPaths, ...acceptance];
  const context = items
    .filter((item): item is string => typeof item === "string" && Boolean(item.trim()))
    .map((item) => item.trim())
    .join(" ");

  if (R4.test(context)) {
    return {
      riskClass: "R4",
      reasons: ["systemic or architecture-level signal detected"],
      explicit: false,
    };
  }
  if (R3.test(context)) {
    return {
      riskClass: "R3",
      reasons: ["critical security, data, production, or concurrency signal detected"],
      explicit: false,
    };
  }
  if (R2.test(context)) {
    return {
      riskClass: "R2",
      reasons: ["integration or public-contract signal detected"],
      explicit: false,
    };
  }
  if (R0.test(context) && !CHANGE.test(context)) {
    return {
      riskClass: "R0",
      reasons: ["mechanical or documentation-only signal detected"],
      explicit: false,
    };
  }
  const codePaths = allowedPaths.some(
    (path) => path.startsWith("src/") || path.startsWith("app/") || path.startsWith("packages/"),
  );
  if (R0.test(context) && !codePaths) {
    return {
      riskClass: "R0",
      reasons: ["mechanical or documentation-only signal detected"],
      explicit: false,
    };
  }
  return {
    riskClass: defaultRisk,
    reasons: [`no higher-risk signal detected; using default ${defaultRisk}`],
    explicit: false,
  };
}
